evmrpc/filter.go

```diff
@@ -0,0 +1,268 @@
+package evmrpc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/filters"
+	abci "github.com/tendermint/tendermint/abci/types"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+)
+
+type filter struct {
+	fc       filters.FilterCriteria
+	deadline *time.Timer
+	cursors  map[common.Address]string
+}
+
+type FilterAPI struct {
+	tmClient     rpcclient.Client
+	nextFilterID uint64
+	filtersMu    sync.Mutex
+	filters      map[uint64]filter
+	filterConfig *FilterConfig
+}
+
+type FilterConfig struct {
+	timeout time.Duration
+}
+
+func NewFilterAPI(tmClient rpcclient.Client, filterConfig *FilterConfig) *FilterAPI {
+	filters := make(map[uint64]filter)
+	api := &FilterAPI{
+		tmClient:     tmClient,
+		nextFilterID: 1,
+		filtersMu:    sync.Mutex{},
+		filters:      filters,
+		filterConfig: filterConfig,
+	}
+
+	go api.timeoutLoop(filterConfig.timeout)
+
+	return api
+}
+
+func (a *FilterAPI) timeoutLoop(timeout time.Duration) {
+	ticker := time.NewTicker(timeout)
+	defer ticker.Stop()
+	for {
+		<-ticker.C
+		a.filtersMu.Lock()
+		for id, filter := range a.filters {
```

Comment:
will the ordering of iteration here affect state? wondering b/c it could be nondeterministic cc @codchen

Reply:
this is transient state in server, not stored as chain state